plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -41,20 +42,47 @@ abstractly specifiable. It's an implementation detail.
 
 -}
 
+{- Note [Integer types for costing]
+Using machine integers rather than arbitrary-sized 'Integer's gives us a nice speedup.
+
+But we also need to care about overflow: the cost counters overflowing is a potential attack!
+
+So we use a variant of 'Data.SafeInt' that does saturating arithmetic (i.e. returns max/min bound
+on overflow instead of throwing), which works well for us since we're only doing monotonic operations,
+and nobody is going to set a budget anywhere near 'maxBound'. Also we don't need to deal with exceptions,
+which is nice.
+
+This isn't quite as fast as using 'Int64' directly (I don't know why that would be, apart from maybe
+just the overflow checks), but we need the safety.
+
+One other wrinkle is that 'SatInt' is backed by an 'Int' (i.e. a machine integer with platform-dependent
+size), rather than an 'Int64' since the primops that we need are only available for 'Int' until GHC 9.2
+or so.
+
+This is okay, because we don't expect budgets to be anywhere near 'maxBound' for even Int32! So we're
+never going to need to worry about the possibility of saturating for 'Int32' but not for 'Int64' or something.
+-}
+
 -- | Counts size in machine words (64bit for the near future)
-newtype ExMemory = ExMemory Integer
+newtype ExMemory = ExMemory SatInt
   deriving (Eq, Ord, Show)
-  deriving newtype (Num, Pretty, NFData)
-  deriving (Semigroup, Monoid) via (Sum Integer)
-deriving newtype instance PrettyDefaultBy config Integer => PrettyBy config ExMemory
+  deriving newtype (Num, NFData)
+  deriving (Semigroup, Monoid) via (Sum SatInt)
+instance Pretty ExMemory where
+    pretty (ExMemory i) = pretty (toInteger i)
+instance PrettyDefaultBy config SatInt => PrettyBy config ExMemory where
+    prettyBy _ m = pretty m
 
 -- TODO: 'Integer's are not particularly fast. Should we use @Int64@?
 -- | Counts CPU units - no fixed base, proportional.
-newtype ExCPU = ExCPU Integer
+newtype ExCPU = ExCPU SatInt
   deriving (Eq, Ord, Show)
-  deriving newtype (Num, Pretty, NFData)
-  deriving (Semigroup, Monoid) via (Sum Integer)
-deriving newtype instance PrettyDefaultBy config Integer => PrettyBy config ExCPU
+  deriving newtype (Num, NFData)
+  deriving (Semigroup, Monoid) via (Sum SatInt)
+instance Pretty ExCPU where
+    pretty (ExCPU i) = pretty (toInteger i)
+instance PrettyDefaultBy config SatInt => PrettyBy config ExCPU where
+    prettyBy _ m = pretty m
```

Comment:
This is a bit weird: you require `SatInt` to be printable via `config`, but ignore the config at the term level. You can either drop the constraint or just use the old `deriving newtype` and replace `Integer` with `SatInt`.